RainbowToasts

natew opened the pull request:
Fixes APP-2818

## What changed (plus any additional context for devs)

Implementing new designs.

## Screen recordings / screenshots


## What to test

natew commented on src/components/rainbow-toast/MintToastIcon.tsx:
```diff
@@ -0,0 +1,29 @@
+    <View
+      style={{
+        backgroundColor: colors.background,
+        borderRadius: 10,
```
squircle this?

natew commented on src/components/rainbow-toast/RainbowToast.tsx:
Took a bit to get both to play together right so leaving this, whats nice about it now is it does both at once so as you press + swipe it pushes in a bit which imo looks good.

mikedemarais replied:
i agree supporting "press + swipe" is nice and worth doing, but our existing GestureHandlerButton allows you to configure simultaneous gesture recognizers so you should be able to still achieve that using our existing component instead of having all this duplicated button press related code in this file

maxbbb reviewed (changes requested):
Had some Issues in testing via the dev menu. Not sure if these are specific to them being mock transactions or if they are real issues.
  	- Pressing the add dapp swap transaction adds a pending transaction but no toast displays
  	- Adding three of any of the same transaction results in the third not displaying. 
  	- I experienced both of these, and nothing changed since, but I can no longer reproduce either of them. Listing in case it's helpful. 
  		- When I click any of the "add" transaction types in the dev menu twice, the first one immediately moves to the success state. It gets dismissed but never removed from the pending transactions
  		- Every other time after doing the following sequence - clear all -> add send transaction -> update send to failed - I would get the UI for either failed or a success, not consistently failed. 

General comments
  - Try to use Stylesheet where possible or a memoized style object rather than inline objects.

natew commented on src/components/rainbow-toast/ToastExpandedContent.tsx:
```diff
@@ -0,0 +1,197 @@
+
+  return (
+    <View style={{ maxWidth: '33%', flexDirection: 'column', minHeight: '100%', gap: 12, marginVertical: -4 }}>
+      <Text ellipsizeMode="tail" numberOfLines={1} color={{ custom: colors.foregroundDim }} size="13pt" weight="medium">
+        {topValue || 'Pending'}
+      </Text>
+      <Text ellipsizeMode="tail" numberOfLines={1} color={{ custom: colors.foreground }} size="15pt" weight="medium" align="right">
+        {bottomValueWithoutSymbol || '0 ETH'}
+      </Text>
```
Yes that's due to the mock data, in real data it shows the right amounts, it's because it pulls it from the app live data which pending transactions don't have.

mikedemarais replied:
why create mock data with bad values though? why not fix the mock data to reflect the real data?

natew replied:
The mock data has valid values of past transactions, but in order to make the debug panel work fully like real toasts I’d have to modify other areas of the codebase so they pass through the rest of the stores as though they were real. 

I didn’t look into doing that because the debug panel here served its purpose - letting me test all the states of the toasts much more quickly and thoroughly than manually creating transactions. The only thing it misses by not moving transactions all the way through is the right-side detail in the expanded state, which shows fallback values to at least see the formatting. 

I think to served its purpose well, it’s a nice dev tool for us to use in the future. Now that I’m done i can move it into the other dev tools area though.

natew commented on src/components/rainbow-toast/ToastSFSymbolIcon.tsx:
```diff
@@ -0,0 +1,67 @@
+import React from 'react';
+import { StyleSheet, Text, View } from 'react-native';
+
+const sfSymbols = {
+  check: IS_IOS ? '􀆅' : <Icon style={{ marginLeft: 6, marginTop: 6 }} width={18} height={18} color="white" name="checkmark" />,
+  exclamationMark: IS_IOS ? '􀅎' : <Icon size={12} color="white" name="warning" />,
+};
```
They weren't showing for me, one was a box and other was a weird mark. Not sure why.

maxbbb replied:
Interesting, did you see this on a physical device as well? Either way we should figure out how to get them to work or pick different icons rather than branching platform logic

natew commented on src/components/rainbow-toast/getToastFromTransaction.ts:
```diff
@@ -0,0 +1,131 @@
+const txIdToToastId = (tx: RainbowTransaction) => tx.hash + (tx.address || tx.asset?.address) + (tx.chainId || tx.asset?.chainId);
+
+export function getToastFromTransaction(tx: RainbowTransaction, mints?: Mints): RainbowToast | null {
+  if (tx.swap || tx.type === 'swap') {
```
When originally building it I did have a reason for this, testing data I saw it was swap. I'll go back and check and see if I can cause it to not be swap.

natew replied:
Checked a couple swaps and seems it's not needed.

natew commented on src/components/rainbow-toast/ToastSFSymbolIcon.tsx:
```diff
@@ -0,0 +1,67 @@
+          StyleSheet.absoluteFillObject,
+          {
+            backgroundColor: bg,
+            borderRadius: borderRadius - 5,
```
adding a comment explaining

natew commented on src/components/rainbow-toast/RainbowToast.tsx:
```diff
@@ -0,0 +1,398 @@
+        translateX.value = event.translationX;
+      })
+      .onChange(event => {
+        // at least on iOS simulator velocityX is always 0 so using this
+        lastChangeX.value = event.changeX;
```
Looks like this bug is fixed in newer RN versions:

https://github.com/software-mansion/react-native-reanimated/issues/3758

I'm hesitant to remove it until we upgrade to 79 - in part because I am much faster developing in the simulator so I enjoy having everything work there, and also I'd assume that anyone else testing this out on the simulator would come to me and ask if it was a bug.

I'll see about detecting simulator and only fixing it there, and make a follow-on PR for after the 0.79 merge that removes it.

natew commented on src/components/rainbow-toast/ToastContent.tsx:
```diff
@@ -0,0 +1,154 @@
+
+export const getSwapToastNetworkLabel = ({ toast }: { toast: RainbowToastSwap }) => {
+  return (
+    <RNText>
+      {toast.fromAssetSymbol} <RNText style={{ fontWeight: '200' }}>􀄫</RNText> {toast.toAssetSymbol}
+    </RNText>
```
Because the `ToastContentDisplay` controls the color/size so RNText will inherit both without me having to re-define it, but still needed to control the weight on the inner -> icon to be lighter

maxbbb commented on src/components/rainbow-toast/RainbowToast.tsx:
```diff
@@ -0,0 +1,398 @@
+      .onEnd(() => {
+        runOnJS(showExpanded)();
+      });
+  }, [isPressed, showExpanded, touchStartedAt, translateX.value]);
```
What I'm seeing on my iOS device is if you press and hold the toast for longer than 2s (don't pan around as that will cancel the tap before the 2s) it gets stuck in the pressed state with the toast stuck scaled down.

natew replied:
With my latest push that made maxDuration really long? I don't see that.

maxbbb replied:
The reason I said in my original point that "it's not as simple as extending the maxDuration" is because now the press action triggers no matter how long you press down for. Before, when the maxDuration was just two seconds, what would happen is if you pressed and held the toast would get stuck in the scaled in state. The new behavior I think is preferred compared to the toast getting stuck, but still unideal. However this definitely isn't a blocker, can revisit some other time.

natew replied:
Fixed this in latest

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/nate/toasts/15d861dc3ea642ee8ebe3ec158fedd26d7147086.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/nate/toasts/15d861dc3ea642ee8ebe3ec158fedd26d7147086.ipa) for 15d861dc3ea642ee8ebe3ec158fedd26d7147086

natew commented:
@maxbbb made the following changes:

- Android icons now using SF font, not sure what happened here but testing on a real device they worked and after rebuilding the emulator they were loading correctly
- Added logic to handle wider toasts, for now it's actually simple since we know the widths. In the simulator it drops frames but on my real device it looks smooth
- Moved the debug tools into the existing dev tools area
- Added worklets in various places and tested
- Cleaned up naming
- Using both real android and iOS improved some of the thresholds to feel good
- DApp transactions show a loading spinner now like the others in expanded state

I believe this resolves all the remaining issues, minus the conversations in the last unresolved threads above which I've replied to.

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/nate/toasts/65a27a793c623836af3b621fa444205fc02f2229.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/nate/toasts/65a27a793c623836af3b621fa444205fc02f2229.ipa) for 65a27a793c623836af3b621fa444205fc02f2229

maxbbb commented on src/components/rainbow-toast/ToastSFSymbolIcon.tsx:
```diff
@@ -0,0 +1,71 @@
+            color={{ custom: colors.white }}
+            weight="bold"
+            // somehow this does make a difference, 0.25 specifically vs 0.5 in centering it right
+            style={{ transform: [{ translateX: 0.25 }] }}
```
The reason is because the smallest representable unit on most devices is 1/3 pixel. Typically you want to use Stylesheet.hairlineWidth to represent that unit. Some SF icons are a little hard to center properly and there is a specific <TextIcon> component and size="icon 13px" prop that might help not have to have this transform here, but not sure.

maxbbb commented on src/components/rainbow-toast/RainbowToastExpandedDisplay.tsx:
Spoke with Christian about this. Current interaction is fine, but if we could please use the existing Panel component and backdrop (no blur) so that the design is consistent with others of this type and not duplicated. Also where possible use one of the animation configs that's in the `animations/animationConfigs` so that they feel consistent with others in the app.

The other thing is just some cleanup on the spacing and radius to match the designs. 

App:
<img width="613" height="433" alt="Screenshot 2025-07-29 at 11 53 49 AM" src="https://github.com/user-attachments/assets/df0e6b3e-a077-476d-995c-76aaa065d6a2" />

Designs:

 
<img width="312" height="211" alt="Screenshot 2025-07-29 at 11 54 07 AM" src="https://github.com/user-attachments/assets/42d4b85b-6f92-48b2-8041-9e9a65aec14d" />

natew replied:
Adjusted to use Panel + backdrop settings, adjusted spacing, though have to account for swap icon being wider and not in designs so it's a balancing act, Christian mentioned we can make the icons a bit wider which I've done:

<img width="882" height="646" alt="CleanShot 2025-07-29 at 09 37 41@2x" src="https://github.com/user-attachments/assets/5503b335-e7f5-4200-9d22-dfc29390b34b" />

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/nate/toasts/eaabffd72efc760aab0e73b8d582af9c0505e0d0.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/nate/toasts/eaabffd72efc760aab0e73b8d582af9c0505e0d0.ipa) for eaabffd72efc760aab0e73b8d582af9c0505e0d0

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/nate/toasts/c3d0c27ab5c9f25c6b463a67e9e5334dc071f367.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/nate/toasts/c3d0c27ab5c9f25c6b463a67e9e5334dc071f367.ipa) for c3d0c27ab5c9f25c6b463a67e9e5334dc071f367

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/nate/toasts/0d32edeb808874bf1990ac6160022ed8ec8cb91b.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/nate/toasts/0d32edeb808874bf1990ac6160022ed8ec8cb91b.ipa) for 0d32edeb808874bf1990ac6160022ed8ec8cb91b

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/nate/toasts/d57738eaf6c44ca4216d92c30c1e91f59f84cf94.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/nate/toasts/d57738eaf6c44ca4216d92c30c1e91f59f84cf94.ipa) for d57738eaf6c44ca4216d92c30c1e91f59f84cf94

jinchung commented on src/components/rainbow-toast/ToastDebugSheet.tsx:
```diff
@@ -0,0 +1,269 @@
+/* eslint-disable no-plusplus */
+
+import {
```
actually I would prefer if we removed `ToastDebugSheet` and the changes related to `isMocked`

jinchung commented on src/screens/SendSheet.tsx:
```diff
@@ -669,9 +672,14 @@ export default function SendSheet() {
+        // without toasts: navigate user to the activity list
+        if (!rainbowToastsEnabled) {
+          InteractionManager.runAfterInteractions(() => {
+            navigate(Routes.PROFILE_SCREEN);
```
there are a couple other places where we trigger this navigation behavior after a transaction, outside of just the send sheet, e.g, NFTSingleOfferSheet/index.ts, MintSheet.tsx (we can ignore updating ENSConfirmRegisterSheet.tsx as we'll likely be ripping that out soon)

jinchung commented on src/screens/token-launcher/components/TokenLogo.tsx:
```diff
@@ -1,4 +1,4 @@
-import React, { useCallback, useMemo } from 'react';
+import React, { useCallback, useEffect, useMemo } from 'react';
```
not needed

jinchung commented on src/components/rainbow-toast/getToastFromTransaction.ts:
```diff
@@ -0,0 +1,104 @@
+import Routes from '@/navigation/routesNames';
+import { Mints } from '@/resources/mints';
+
+export const txIdToToastId = (tx: RainbowTransaction): string => (tx.nonce ? `${tx.nonce}` : tx.hash + (tx.chainId || tx.asset?.chainId));
```
leaving a comment to revisit this - we need to handle scenarios of switching wallets

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/nate/toasts/7832a79ab3fb05f6b0e4917f4b217a1a6106adf0.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/nate/toasts/7832a79ab3fb05f6b0e4917f4b217a1a6106adf0.ipa) for 7832a79ab3fb05f6b0e4917f4b217a1a6106adf0

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/nate/toasts/fbb402a06d4c74771e8d03a30b26434f1a6d55db.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/nate/toasts/fbb402a06d4c74771e8d03a30b26434f1a6d55db.ipa) for fbb402a06d4c74771e8d03a30b26434f1a6d55db

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/nate/toasts/40bf9de85a2521a5ea9d8662ebef048bf7e1f851.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/nate/toasts/40bf9de85a2521a5ea9d8662ebef048bf7e1f851.ipa) for 40bf9de85a2521a5ea9d8662ebef048bf7e1f851

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/nate/toasts/23b67b1bfbbe25db1a5fb05644dd4ff13f309947.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/nate/toasts/23b67b1bfbbe25db1a5fb05644dd4ff13f309947.ipa) for 23b67b1bfbbe25db1a5fb05644dd4ff13f309947

ibrahimtaveras00 reviewed (approved):
All issues have been resolved, QA Passed 👍🏽
